src/main/java/com/mongodb/spark/sql/connector/write/MongoDataWriter.java

```diff
@@ -16,96 +16,181 @@
+    this.writeConfig = writeConfig;
+    this.epochId = epochId;
+    this.bulkWriteOptions = new BulkWriteOptions().ordered(writeConfig.isOrdered());
+  }
 
   /**
-   * Commits this writer after all records are written successfully, returns a commit message which
-   * will be sent back to driver side and passed to {@link
-   * MongoBatchWrite#commit(WriterCommitMessage[])}.
-   *
-   * <p>The written data should only be visible to data source readers after {@link
-   * MongoBatchWrite#commit(WriterCommitMessage[])} succeeds, which means this method should still
-   * "hide" the written data and ask the {@link MongoBatchWrite} at driver side to do the final
-   * commit via {@link WriterCommitMessage}.
+   * Converts the record into a {@link BsonDocument} and stages the write.
    *
-   * <p>If this method fails (by throwing an exception), {@link #abort()} will be called and this
-   * data writer is considered to have been failed.
+   * <p>Once {@link WriteConfig#getMaxBatchSize} is hit then the bulk operation takes place.
    *
-   * @throws IOException if failure happens during disk/network IO like writing files.
+   * @param record the row to be written
+   * @see WriteConfig#getMaxBatchSize
    */
   @Override
-  public WriterCommitMessage commit() throws IOException {
-    return null;
+  public void write(final InternalRow record) {
+    BsonDocument bsonDocument = rowToBsonDocumentConverter.fromRow(record);
+    writeModelList.add(getWriteModel(bsonDocument));
+    if (writeModelList.size() >= writeConfig.getMaxBatchSize()) {
+      writeModels();
+    }
   }
 
   /**
-   * Aborts this writer if it is failed. Implementations should clean up the data for already
-   * written records.
+   * Commits this writer after all records are written successfully.
    *
-   * <p>This method will only be called if there is one record failed to write, or {@link #commit()}
-   * failed.
+   * <p>Ensures any remain writes in the batch are written.
    *
-   * <p>If this method fails(by throwing an exception), the underlying data source may have garbage
-   * that need to be cleaned by {@link MongoBatchWrite#abort(WriterCommitMessage[])} or manually,
-   * but these garbage should not be visible to data source readers.
+   * @return a MongoWriterCommitMessage
+   */
+  @Override
+  public WriterCommitMessage commit() {
+    writeModels();
+    LOGGER.debug("Finished all writes for: PartitionId: {}, TaskId: {}.", partitionId, taskId);
+    return new MongoWriterCommitMessage(partitionId, taskId, epochId);
+  }
+
+  /**
+   * Aborts the write.
    *
-   * @throws IOException if failure happens during disk/network IO like writing files.
+   * <p>Note: Data is not cleaned up and will require manual cleaning.
    */
   @Override
-  public void abort() throws IOException {}
+  public void abort() {
+    LOGGER.debug("Aborting write for: PartitionId: {}, TaskId: {}.", partitionId, taskId);
+    releaseClient();
+  }
```

Comment:
I think that similarly to {{MongoBatchWrite.abort}}, this method must complete abruptly to express that the data was not rolled back. The specification of the method {{DataWriter.abort}} says _"If this method fails(by throwing an exception), the underlying data source may have garbage"_.

Reply:
Sounds good - we'll still have to document in DOCSP-19782 that the data will be visible to a datasource reader.